File: renamer.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
#
###############################################################################
# Copyright (c) 2010-2011, Gianluca Fiore <forod.g@gmail.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
###############################################################################

__author__ = "Gianluca Fiore"
__license__ = "GPL"
__version__ = "0.2"
__date__ = "18/11/2010"
__email__ = "forod.g@gmail.com"

import os
import re
import sys
import datetime
import argparse




def argument_parser():
    argparser = argparse.ArgumentParser()
    argparser.add_argument("-s", "--no-spaces",
            help="substitute spaces",
            action="store_true",
            dest="substitute_spaces")
    argparser.add_argument("-a", "--append",
            help="Append string to the end of filenames",
            action="store",
            type=str,
            dest="append_string")
    argparser.add_argument("-p", "--prefix",
            help="Prefix string to filenames",
            action="store",
            type=str,
            dest="prefix_string")
    argparser.add_argument("-r", "--remove",
            help="Remove a string",
            action="store",
            type=str,
            dest="remove_string")
    argparser.add_argument("-m", "--minimize",
            help="Minimize the extension",
            action="store_true",
            dest="minimize_extension")
    argparser.add_argument("-t", "--translate",
            help="Substitute (translate) a series of characters with another",
            action="store",
            dest="translate",
            type=str,
            nargs=2)
    argparser.add_argument("-n", "--numbering",
            help="Rename files adding a numeration",
            action="store",
            type=str,
            dest="numbering",
            nargs=2)
    argparser.add_argument("-d", "--date",
            help="Prepend date to files",
            action="store",
            type=str,
            dest="date_fmt")
    argparser.add_argument("-g", "--gui",
            help="Start the GUI",
            action="store_true",
            dest="gui_enable")
    argparser.add_argument(action="store",
            help="Files",
            dest="files",
            nargs="+")
    options = argparser.parse_args()
    return options


def cbk_translate(option, opt_str, value, parser):
    """Callback function for translate"""
    value = [] 

    for arg in parser.rargs:
        if arg[:2] == "--" and len(arg) > 2:
            break
        if arg[:1] == "-" and len(arg) > 1:
            break
        value.append(arg)

        del parser.rargs[:len(value)]
        setattr(parser.values, option.dest, value)

def check_length(typ, length):
    """Check that the length for the given type is exactly as it should 
    (length)"""
    if len(typ) != length:
        raise argparse.ArgumentError("You need to give exactly %d arguments" % length)
    else:
        return True

def ispath(path):
    """Exit if path is not a correct path"""
    if os.path.isdir(path):
        return True
    else:
        sys.stderr.write("You must give a valid path")
        sys.exit(1)

def filelist(*args):
    """Given or a valid path or a number of files (or both), return a list
    of filenames in the given directory or comprising the files as arguments"""
    filenames = []
    for a in args:
        for i in a:
            # set some variables about the file
            bname = os.path.basename(i)
            dname = os.path.dirname(i)
            root = os.path.abspath(dname)
            completename = root + '/' + bname
            if os.path.isfile(completename):
                filenames.append(completename)
            elif ispath(i):
                for files in os.listdir(i):
                    fl = os.path.join(root, files)
                    if os.path.isfile(fl):
                        filenames.append(fl)
            else:
                print("Not a path and not a file, check the list of arguments")

        #fll = [fille for fille in os.listdir(path) if os.path.isfile(os.path.join(root, files))]
    return filenames

def sub_spaces(filenames):
    """Substitute all spaces in filenames with underscores"""
    for files in filenames:
        newname = re.sub(' ', '_', files)
        os.rename(files, newname)

def append_str(filenames, s):
    """Append given string to the end of all filenames"""
    for files in filenames:
        root = os.path.dirname(files)
        oldname = os.path.basename(files)
        newname = root + '/' + oldname + s
        os.rename(files, newname)

def prefix_str(filenames, s):
    """Prefix all filenames with the given string"""
    for files in filenames:
        root = os.path.dirname(files)
        oldname = os.path.basename(files)
        newname = root + '/' + s + oldname
        os.rename(files, newname)

def remove_str(filenames, s):
    """Remove a given string from all filenames"""
    for files in filenames:
        root = os.path.dirname(files)
        oldname = os.path.basename(files)
        if re.search(s, oldname):
            # it's a regexp
            newname = root + '/' + re.sub(s, '', oldname)
        else:
            # not a regexp, simple replace
            newname = root + '/' + oldname.replace(s, '')
        try:
            os.rename(files, newname)
        except OSError:
            # it's possible that the regexp might trying to cancel all the
            # characters in filename. Skip that file
            continue

def minimize_ext(filenames):
    """Minimize all file extensions of given files"""
    for files in filenames:
        name, ext = os.path.splitext(files)
        os.rename(files, name + ext.lower())

def translate_chars(filenames, char_f, char_t):
    """Translate every given characters in filenames to another character"""
    for files in filenames:
        root = os.path.dirname(files)
        oldname = os.path.basename(files)
        newname = root + '/' + re.sub(char_f, char_t, oldname)
        os.rename(files, newname)

def idx_numbering(filenames, name, int_start):
    """Rename every files with the given name and adding a numerations starting 
    at the given number"""
    try:
        n = int(int_start)
    except ValueError:
        # perhaps we forgot to give the int to start numbering, defaulting to 1
        n = 1
    # filenames must be sorted beforehand
    for files in sorted(filenames):
        root = os.path.dirname(files)
        oldname, ext = os.path.splitext(files)
        # prefix enough 0 if the number is not already at least 3 digits long
        if n < 10:
            idx = '00' + str(n)
        elif n >= 10 and n < 100:
            idx = '0' + str(n)
        else:
            idx = str(n)
        newname = root + '/' + name + idx + ext
        n += 1
        os.rename(files, newname)

def prepend_date(filenames, date_fmt):
    """Prepend each files with the current date"""
    today = datetime.date.today()
    print("If you want to use a date different from today's, append it directly as a string")
    for files in filenames:
        root = os.path.dirname(files)
        oldname = os.path.basename(files)
        newname = root + '/' + today.strftime(date_fmt) + '-' + oldname
        os.rename(files, newname)


#class RenamerWindow(QtGui.QMainWindow):
#    def __init__(self, parent=None):
#        QtGui.QWidget.__init__(self, parent)
#
#
#        self.setGeometry(300, 300, 450, 350)
#        self.setWindowTitle('Renamer')
#
#
#        exit_icon = QtGui.QAction(QtGui.QIcon('/usr/share/icons/gnome/24x24/actions/exit.png'), 'Exit', self)
#        exit_icon.setShortcut('Ctrl+Q')
#
#        preferences = QtGui.QAction(QtGui.QIcon('/usr/share/icons/gnome/24x24/categories/preferences-other.png'), 'Preferences', self)
#        preferences.setShortcut('Ctrl+P')
#
#        filedialog = QtGui.QAction(QtGui.QIcon('/usr/share/icons/gnome/24x24/actions/fileopen.png'), 'Open File', self)
#        filedialog.setText('Open File')
#        self.connect(filedialog, QtCore.SIGNAL('activated()'), QtCore.SLOT(QtGui.QFileDialog.getOpenFileName("", "*.py", self, "FileDialog")))
#
#        self.connect(exit_icon, QtCore.SIGNAL('triggered()'), QtCore.SLOT('close()'))
#
#        menubar = self.menuBar()
#        options_menu = menubar.addMenu('&Options')
#        options_menu.addAction(preferences)
#        options_menu.addAction(exit_icon)







def main():
    options = argument_parser()

    # do we want a gui?
    if options.gui_enable:
        try:
            from PyQt4 import QtGui, QtCore
        except:
            sys.exit(1)

            app = QtGui.QApplication(sys.argv)

            w = RenamerWindow()
            w.show()
            app.exec_()
            return

    # get all the filenames
    filenames = filelist(options.files)
    if options.substitute_spaces:
        sub_spaces(filenames)
    elif options.append_string:
        append_str(filenames, options.append_string)
    elif options.prefix_string:
        prefix_str(filenames, options.prefix_string)
    elif options.remove_string:
        remove_str(filenames, options.remove_string)
    elif options.minimize_extension:
        minimize_ext(filenames)
    elif options.translate:
        check_length(options.translate, 2)
        translate_chars(filenames, options.translate[0], options.translate[1])
    elif options.numbering:
        check_length(options.numbering, 2)
        idx_numbering(filenames, options.numbering[0], options.numbering[1])
    elif options.date_fmt:
        prepend_date(filenames, options.date_fmt)
    else:
        print("What do you exactly want to do?")

        

if __name__ == '__main__':
    status = main()
    sys.exit(status)

File: renamer.lua
#!/usr/bin/env lua

---
-- @author Gianluca Fiore
-- @copyright 2012-2013, Gianluca Fiore <forod.g@gmail.com>
--

local lfs = require("lfs")

local _O = {}

function print_help()
	print([[Wrong arguments

USAGE:
	rename.lua [-h][-s|--substitute][-a|--append=APPEND_STRING][-p|--prefix=PREFIX_STRING]
		[-r|--remove=REMOVE_STRING][-m|--minimize][-t|--translate=TRANSLATE_FROM,TRANSLATE_TO]
		[-n|--numbering=NUMBERING,NAME][-d|--date=DATE_FMT][-D|--no-dashes] 
		files|directory [files|directories ...]

Arguments:
	-h					this help
	-s, --substitute			substitute spaces with underscores
	-a, --append=<string>			append a string to filenames
	-p, --prefix=<string>			prefix a string to filenames
	-r, --remove=<string>			remove <string> from filenames
	-m, --minimize				minimize extensions from filenames
	-t, --translate=<from>,<to>		translate all <from> characters to <to> characters in filenames
	-n, --numbering=<numbering>,<name>	rename all filenames as <name> with a progressive numeration starting at <numbering>
	-d, --date=<date>			add a date to the beginning of filenames (use [today|yesterday|tomorrow|number])
	-D, --no-dashes				strip dashes from dates. Only meaningful with -d, --date
		]])
end

function print_date_help()
	print([[Incorrect/missing arguments

USAGE
	today			for today's date
	yesterday		for yesterday's date
	tomorrow		for tomorrow's date
	year-month-day		for the exact date (as in '2012-10-01')
	a single number		for a specific date of current the month ('20' will become '2012-10-20'),
				or only the year as it is (if the number is biggest than 31, which is always
				considered a day then)

all dates will be in the "yyyy-mm-dd" format unless the -D or --no-dashes options was given; in this case, dashes are removed]])
end


function cli_parse(...)
	-- this is probably too long/convoluted. I didn't want to use 
	-- external libraries so WillFixThisLater™
	local args = ...
	local files = {}
	for i = #args, 1, -1 do
		local a = arg[i]
		if a:match("^%-h") then
			print_help()
			os.exit(0)
		elseif a:match ("^[-]+(.*)") then
			if a:match("^%-s") or a:match("^[-]+substitute") then
				_O.substitute = true
			elseif a:match("^%-m") or a:match("^[-]+minimize") then
				_O.minimize = true
			elseif a:match("^%-a") or a:match("^[-]+append") then
				_O.append = true
				local var, val = a:match("([a-z_%-]*)=(.*)")
				if val then
					_O.append_string = val
				else
					-- no flag=string pattern? Then assume -flag string 
					-- pattern. Same thing in the others occurrences
					_O.append_string = arg[i+1]
					i = i + 1
				end
			elseif a:match("^%-p") or a:match("^[-]+prefix") then
				_O.prefix = true
				local var, val = a:match("([a-z_%-]*)=(.*)")
				if val then
					_O.prefix_string = val
				else
					_O.prefix_string = arg[i+1]
					i = i + 1
				end
			elseif a:match("^%-r") or a:match("^[-]+remove") then
				_O.remove = true
				local var, val = a:match("([a-z_%-]*)=(.*)")
				if val then
					_O.remove_string = val
				else
					_O.remove_string = arg[i+1]
					i = i + 1
				end
			elseif a:match("^%-t") or a:match("^[-]+translate") then
				_O.translate = true
				local var, val_f, val_t = a:match("([a-z_%-]*)=(.*),(.*)")
				if val_f and val_t then
					_O.translate_from, _O.translate_to = val_f, val_t
				else
					_O.translate_from, _O.translate_to = arg[i+1], arg[i+2]
					i = i + 2
				end
			elseif a:match("^%-n") or a:match("^[-]+numbering") then
				_O.numbering = true
				local var, idx, name = a:match("([a-z_%-]*)=(.*),(.*)")
				if idx and name then
					-- check which between idx and name is the number 
					-- and which is the string. The user may bave 
					-- inverted them
					if type(idx) == "number" or type(tonumber(idx)) == "number" then
						_O.numbering_idx = idx
						_O.numbering_name = name
					elseif type(idx) == "string" and type(tonumber(name)) == "number" then
						_O.numbering_idx = name
						_O.numbering_name = idx
					else
						if tonumber(idx) == nil then
							-- impossible to have a number and a string, 
							-- bailing out
							print_help()
							os.exit(1)
						end
						-- just try to have a number and a string
						_O.numbering_idx = tonumber(idx)
						_O.numbering_name = tostring(name)
					end
				else
					-- if -t string string pattern, try which string can 
					-- be the index
					if tonumber(arg[i+1]) then
						_O.numbering_idx = arg[i+1]
						_O.numbering_name = tostring(arg[i+2])
						i = i + 2
					elseif tonumber(arg[i+2]) then
						_O.numbering_idx = arg[i+2]
						_O.numbering_name = tostring(arg[i+1])
						i = i + 2
					else
						-- if all else fails, first string is the index 
						-- and the second the name.
						-- Blame on you user for not reading the help
						_O.numbering_idx = arg[i+1]
						_O.numbering_name = tostring(arg[i+2])
						i = i + 2
					end
				end
			elseif a:match("^%-d") or a:match("^[-]+date") then
				_O.date = true
				local var, val = a:match("([a-z_%-]*)=(.*)")
				if val then
					_O.date_input = val
				else
					_O.date_input = arg[i+1]
					i = i + 1
				end
			elseif a:match("^%-D") or a:match("^[-]+no[-]dashes") then
				_O.dashes = false
			else
				print_help()
				os.exit(1)
			end
		elseif a:match("^([^-][%w_%p]*)") then
			table.insert(files, a)
		else
			print_help()
		end
	end
	return files
end

---Check that the length for the given type is exactly as it should
--@param typ the type
--@param length the desired minimum length
function check_length(typ, length)
	if #typ < length then
		print("Wrong number of arguments, you need to give at least " .. length .. " arguments")
		os.exit(1)
	else
		return true
	end
end

---Extract only the file name from a path string
--@param str the path
function basename(str)
	local name = string.gsub(str, "(.*/)(.*)", "%2")
	return name
end

---Extract only the directory name from a path string
--@param str the path
function dirname(str)
	local name = string.gsub(str, "(.*/)(.*)", "%1")
	return name
end

---Extract only the file extension from a path string. Returns the full
--path and the extension (period included)
--@param str the path
function get_extension(str)
	-- let's check whether the filename seems to have an extension or
	-- not
	if string.match(str, "(.-/)(.-)([.].*)$") then
		local name = string.gsub(str, "(.-/)(.-)([.].*)$", "%1%2")
		local ext = string.gsub(str, "(.-/)(.-)([.].*)$", "%3")
		return name, ext
	else
		local name = string.gsub(str, "(.-/)(.-)([.].*)$", "%1%2")
		return name, ext
	end
end

---Check if a string is a valid path
--@param path the string to check
function ispath(path)
	-- try to rename path to itself; if it works, the original path 
	-- existed
	local s = os.rename(path, path)
	if s == true then
		return true
	else
		return false
	end
end

---Given either a valid path or a number of files (or a mix of both), 
--return a list of filenames in the given directory or comprising the 
--files as arguments
--@param ... the path(s) and/or file(s)
function filelist(...)
	local path = ...
	local files = {}
	for _,e in pairs(path) do
		if lfs.attributes(e, "mode") == "file" then
			table.insert(files, e)
		elseif lfs.attributes(e, "mode") == "directory" then
			for l in lfs.dir(e) do
				if l ~= "." and f ~= ".." then
					local filename = e .. l
					local attr = lfs.attributes(filename)
					if attr.mode == "file" then
						table.insert(files, filename)
					end
				end
			end
		else
			-- neither a file nor a directory? Skip it
--			goto continue
			
		end
		::continue::
	end
	return files
end

---Transform user numerical or alphabetical input in a meaningful date 
--string
--@param input the user input, a string
function dateize(input)
	local date = ''
	if type(input) == "string" then
		if input == "today" then
			date = os.date("%Y-%m-%d")
		elseif input == "yesterday" then
			date = os.date("%Y-%m-%d", os.time()-24*60*60)
		elseif input == "tomorrow" then
			date = os.date("%Y-%m-%d", os.time()+24*60*60)
		elseif string.match(input, "%d+%-%d+%-%d+") then
			date = input
		else
			print_date_help()
			os.exit(1)
		end
	else
		if input <= 31 then
			-- take the input as day of the month and build the date 
			-- string from it
			local m = os.date("%Y-%m")
			date = m .. '-' .. input
		elseif input > 31 then
			-- take input as year and be it the date
			date = input
		else
			-- nothing to do, yet
			print_date_help()
			os.exit(1)
		end
	end
	-- check if the no-dashes cli options has been used
	if _O.dashes == false then
		-- remove dashes from the date string
		date = string.gsub(date, "[-]+", '')
	end
	return date
end

---Substitute all spaces in filename(s) with underscores
--@param filenames a table with the file names
function sub_spaces(filenames)
	for _, f in pairs(filenames) do
		local newname = string.gsub(f, "\\?%s", '_')
		os.rename(f, newname)
	end
end

---Append a string to the end of all filenames given
--@param filenames the files
--@param s the string to append
function append_str(filenames, s)
	local s = tostring(s) -- make sure s is a string
	for _, f in pairs(filenames) do
		local newname = f .. s
		os.rename(f, newname)
	end
end

---Prefix all given filenames with a string
--@param filenames the files
--@param s the string to prefix
function prefix_str(filenames, s)
	local s = tostring(s) -- make sure s is a string
	for _, f in pairs(filenames) do
		local newname = dirname(f) .. s .. basename(f)
		os.rename(f, newname)
	end
end

---Remove a string from all given filenames
--@param filenames the files
--@param s the string to remove
function remove_str(filenames, s)
	local s = tostring(s) -- make sure s is a string
	for _, f in pairs(filenames) do
		local newname = dirname(f) .. string.gsub(basename(f), s, '')
		os.rename(f, newname)
	end
end

---Minimize the extension of given files
--@param filenames the files
function minimize_ext(filenames)
	for _, f in pairs(filenames) do
		local n, e = get_extension(f)
		local newname = n .. string.lower(e)
		os.rename(f, newname)
	end
end

---Translate a list of characters to another in the given filenames
--@param filenames the files
--@param char_f the characters to translate
--@param char_t the characters to translate to
function translate_chars(filenames, char_f, char_t)
	-- have 2 tables to contain each characters in char_f and char_t
	local from_table = {}
	local to_table = {}
	-- make sure characters are strings
	local char_f = tostring(char_f)
	local char_t = tostring(char_t)
	-- limit char_f to the length of char_t. The opposite is already 
	-- done automatically by the lua interpreter
	local correct_char_f = string.sub(char_f, 1, #char_t)
	for _, f in pairs(filenames) do
		local oldname = basename(f) -- save original name
		local newname = oldname -- newname is oldname at the beginning
		-- insert every char in char_f and char_t in corresponding 
		-- tables
		for c in string.gmatch(correct_char_f, ".") do
			table.insert(from_table, c)
		end
		for c in string.gmatch(char_t, ".") do
			table.insert(to_table, c)
		end
		-- iterate over every from_table item and substitute it with the 
		-- same positioned ones in to_table
		for i = 1, #from_table, 1 do
			newname = string.gsub(newname, from_table[i], to_table[i])
		end
		os.rename(f, dirname(f) .. newname)
	end
end

---Rename files with a fixed name and a numbering starting at a given 
--integer
--@param filenames the files
--@param start the integer from which to start the numbering
--@param name the string for the new filename
function idx_numbering(filenames, start, name)
	local idx = tonumber(start)
	for _, f in pairs(filenames) do
		local root = dirname(f)
		local oldname, ext = get_extension(f)
		-- prefix enough 0s if the number is not already at least 3 digits 
		-- long
		if idx < 10 then
			idx = '00' .. idx
		elseif idx >= 10 and idx < 100 then
			idx = '0' .. idx
		else
			idx = idx
		end
		local newname = root .. name .. idx .. ext
		os.rename(f, newname)
		idx = idx + 1
	end
end

---Prepend each file with the current or an arbitrary date
--@param filenames the files
--@param date_fmt the format of the date
function prepend_date(filenames, date_fmt)
	local d = dateize(date_fmt)
	for _,f in pairs(filenames) do
		local root = dirname(f)
		local newname = root .. d .. basename(f)
		os.rename(f, newname)
	end
end


function main()
	local f = cli_parse(arg)
	local fil = filelist(f)
	-- let's see what to do
	if _O.substitute then
		sub_spaces(fil)
	elseif _O.minimize then
		minimize_ext(fil)
	elseif _O.append then
		append_str(fil, _O.append_string)
	elseif _O.prefix then
		prefix_str(fil, _O.prefix_string)
	elseif _O.remove then
		remove_str(fil, _O.remove_string)
	elseif _O.translate then
		translate_chars(fil, _O.translate_from, _O.translate_to)
	elseif _O.numbering then
		idx_numbering(fil, _O.numbering_idx, _O.numbering_name)
	elseif _O.date then
		prepend_date(fil, _O.date_input)
	else
		print_help()
		os.exit(1)
	end
end

main()
